server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -166,6 +166,7 @@ protected void registerRequestHandlers(String actionName, TransportService trans
 
     @Override
     protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        assert request.shardId() != null : "request shardId must be set";
```

Comment:
Interesting - did you run into a situation when this wasn't the case?

Reply:
no, but this work got triggered by looking at open Github issues, in particular https://github.com/elastic/elasticsearch/issues/20279 . In particular I noticed that the desired property to have a complete ShardId when replication requests get to the reroute phase already holds today.
This PR does not close that issue out (needs further investigation on what exactly needs to be done), but addresses the first point on it.